File: web/src/entity/task.ts
import { Resource } from "./resource";
import { UserInfo } from "./user";

export type TaskType = "homework" | "review" | "discussion" | "extra" | "group";
export type TaskPriority = "high" | "medium" | "low";

export interface TaskInfo {
	title: string;
	type: TaskType;
	priority: TaskPriority;
	cost: number;
	deadline: number;
}

export interface Task {
	id: number;
	info: TaskInfo;
	publisher: UserInfo;
}

export const taskTypeText = (str: string): string => {
	switch (str) {
		case "homework":
			return "课程作业";
		case "review":
			return "考试复习";
		case "discussion":
			return "研讨准备";
		case "extra":
			return "竞赛拓展";
		case "group":
			return "小组任务";
		default:
			return str;
	}
};

export const taskPriorityText = (str: string): string => {
	switch (str) {
		case "high":
			return "高";
		case "medium":
			return "中";
		case "low":
			return "低";
		default:
			return str;
	}
};

export interface TaskDetail {
	task: Task;
	description: string;
	resources: Resource[];
	finished: boolean | null;
	myGroup: {
		id: number;
		members: UserInfo[];
	} | null;
	allGroup: TaskGroups | null;
	prevTask: Task | null;
}

export type TaskGroups = {
	[id: number]: { user: UserInfo; finished: boolean }[];
};

File: server/src/entity/task.rs
use serde::{Deserialize, Serialize};
use sqlx::{MySql, Type};
use validator::Validate;

use super::user::UserBasicInfo;

#[derive(PartialEq, Eq, PartialOrd, Ord, Deserialize, Serialize, Clone, Copy)]
#[repr(u32)]
pub enum TaskPriority {
    #[serde(rename = "low")]
    Low = 0,
    #[serde(rename = "medium")]
    Medium = 1,
    #[serde(rename = "high")]
    High = 2,
}

impl Type<MySql> for TaskPriority {
    fn type_info() -> sqlx::mysql::MySqlTypeInfo {
        <u32 as Type<MySql>>::type_info()
    }
}

impl Into<TaskPriority> for u32 {
    fn into(self) -> TaskPriority {
        match self {
            0 => TaskPriority::Low,
            1 => TaskPriority::Medium,
            2 => TaskPriority::High,
            _ => panic!("Invalid task priority"),
        }
    }
}

impl From<TaskPriority> for u32 {
    fn from(value: TaskPriority) -> Self {
        match value {
            TaskPriority::Low => 0,
            TaskPriority::Medium => 1,
            TaskPriority::High => 2,
        }
    }
}

#[derive(Deserialize, Serialize, sqlx::Type, PartialEq, Clone, Copy)]
#[sqlx(type_name = "VARCHAR")]
#[sqlx(rename_all = "lowercase")]
pub enum TaskType {
    #[serde(rename = "homework")]
    Homework,
    #[serde(rename = "review")]
    Review,
    #[serde(rename = "discussion")]
    Discussion,
    #[serde(rename = "extra")]
    Extra,
    #[serde(rename = "group")]
    Group,
}

impl Into<TaskType> for String {
    fn into(self) -> TaskType {
        match self.as_str() {
            "homework" => TaskType::Homework,
            "review" => TaskType::Review,
            "discussion" => TaskType::Discussion,
            "extra" => TaskType::Extra,
            "group" => TaskType::Group,
            _ => panic!("Invalid task type"),
        }
    }
}

#[derive(Deserialize, Serialize, sqlx::FromRow, Validate, Clone)]
pub struct TaskInfo {
    #[validate(length(min = 1))]
    pub title: String,
    #[serde(rename = "type")]
    #[sqlx(rename = "type")]
    pub typ: TaskType,
    pub priority: TaskPriority,
    #[validate(range(min = 1))]
    pub cost: u32,
    pub deadline: u64,
}

#[derive(Deserialize, Serialize, sqlx::FromRow, Clone)]
pub struct Task {
    pub id: u32,
    pub info: TaskInfo,
    pub publisher: u32,
    pub prev: Option<u32>,
}

#[derive(Deserialize, Serialize)]
pub struct TaskDTO {
    pub id: u32,
    pub info: TaskInfo,
    pub publisher: UserBasicInfo,
}
